### PR TITLE
Add flag checker and unique flag generation

### PR DESCRIPTION
**Pull request by zainafzal08**
checker stuff integrated, unsure if me changing the enviorn is gonna break older stuff tho :/

### CODE REVIEW

File: flask_core/app.py [hunk 1]
```diff
@@ -39,4 +55,8 @@ def create_app(config=None):
     # Register all our middleware
     app.wsgi_app = Handler(app.wsgi_app)
 
+    # Set up magic flag generator (ty closure)
+    app.all_flag_ids = app.config["FLAG_IDS"].split(",")
```
Comment (cry) [inline flask_core/app.py]: Why is this here? This is still a config item so it should still live under `app.config`. Run the split in config instead.

File: flask_core/config.py [hunk 2]
```diff
@@ -68,7 +68,8 @@ def __init__(self, **kwargs):
 
         # Try to get user specified config opts, and if they don't exist read from environment
         try:
-            self.FLAG = getattr(self, "FLAG", None) or os.environ["FLAG"]
+            self.FLAG_IDS = getattr(self, "FLAG_IDS", None) or os.environ["FLAG_IDS"]
```
Comment (cry) [inline flask_core/config.py]: Just run the split here.

```python3
self.FLAG_IDS = (getattr(self, "FLAG_IDS", None) or os.environ["FLAG_IDS"]).split(",")
```

File: flask_core/app.py [hunk 3]
```diff
@@ -5,6 +5,22 @@
...
+def gen_flag(app, zid, flag_id):
+    secret = app.config["FLAG_SECRET"]
+    wrapper = app.config["FLAG_WRAP"]
+    s = secret + zid + str(flag_id)
+    b = bytes(s, "utf-8")
+    return f"{wrapper}{{{hashlib.sha256(b).hexdigest()}}}"
+
+
+def check_flag(app, zid, flag):
+    for flag_id in app.all_flag_ids:
```
Comment (cry) [inline flask_core/app.py]:
```python3
return any((app.gen_flag(zid, f) == flag for f in app.config["FLAG_IDS"]))
```

**Review by cry — CHANGES_REQUESTED** [review]

File: flask_core/app.py [hunk 4]
```diff
@@ -39,4 +55,8 @@ def create_app(config=None):
     # Register all our middleware
     app.wsgi_app = Handler(app.wsgi_app)
 
+    # Set up magic flag generator (ty closure)
+    app.all_flag_ids = app.config["FLAG_IDS"].split(",")
+    app.gen_flag = lambda zid, flag_id: gen_flag(app, zid, flag_id)
+    app.check_flag = lambda zid, flag: check_flag(app, zid, flag)
```
Comment (cry) [inline flask_core/app.py]: I forgot, you can use `type.MethodType` here instead.

```python3
app.check_flag = types.MethodType(check_flag, app)
```

**Review by cry — CHANGES_REQUESTED** [review]

File: flask_core/app.py (hunk 4 above)
Comment (zainafzal08) [reply flask_core/app.py]: ok that's awesome god bless

File: flask_core/config.py [hunk 5]
```diff
@@ -68,7 +68,9 @@ def __init__(self, **kwargs):
...
+            self.FLAG_IDS = self.FLAG_IDS.split(",")
```
Comment (cry) [inline flask_core/config.py]: y u no just use my comment's code

**Review by cry — APPROVED** [review]
Ready for merge if you actually tested these changes 💯

File: flask_core/config.py (hunk 5 above)
Comment (zainafzal08) [reply flask_core/config.py]: oh lol missed ur code snippet, will do

File: flask_core/config.py [hunk 6]
```diff
@@ -61,14 +61,20 @@ def __init__(self, **kwargs):
         # Use any user provided config opts
         for k, v in kwargs.items():
             setattr(self, k, v)
-
+        
+        try:
+            self.AUTO_GENERATED_FLAGS = getattr(self, "AUTO_GENERATED_FLAGS", None) or os.environ["FLASK_CORE_AUTO_GENERATED_FLAGS"]
+        except:
```
Comment (cry) [inline flask_core/config.py]: Flake8 will complain here, use `except KeyError`

File: flask_core/app.py [hunk 7]
```diff
@@ -2,12 +2,41 @@
...
+from flask_core.core import bp as core_bp
 from flask_core.middleware.handler import Handler
+import hashlib
+import types
+
+
+def gen_flag(self, zid, flag_id):
+    secret = self.config["FLAG_SECRET"]
+    wrapper = self.config["FLAG_WRAP"]
+    s = secret + zid + str(flag_id)
```
Comment (cry) [inline flask_core/app.py]: y u no hmac

File: flask_core/app.py [hunk 8]
```diff
@@ -47,8 +76,11 @@ def create_app(config=None):
 
     # Register all our middleware
     app.wsgi_app = Handler(app.wsgi_app)
+    # Set up magic flag generator (ty closure)
+    app.gen_flag = types.MethodType(gen_flag, app)
+    app.check_flag = types.MethodType(check_flag, app)
 
+    app.after_request(types.MethodType(grep_flag, app))
```
Comment (cry) [inline flask_core/app.py]: Attach the function directly without `MethodType`, `from flask import current_app`

File: flask_core/config.py [hunk 9]
```diff
@@ -62,13 +62,21 @@ def __init__(self, **kwargs):
         for k, v in kwargs.items():
             setattr(self, k, v)
 
+        try:
+            self.AUTO_GENERATED_FLAGS = (
+                getattr(self, "AUTO_GENERATED_FLAGS", None) or os.environ["FLASK_CORE_AUTO_GENERATED_FLAGS"]
```
Comment (cry) [inline flask_core/config.py]: `os.environ.get("FLASK_CORE_AUTO_GENERATED_FLAGS, True)`

File: flask_core/core/templates/core/home.html [hunk 10]
```diff
@@ -2,4 +2,5 @@
 {% block content %}
 <br>
 <h3>Flask Core common</h3>
+<p>flag{_one}</p>
```
Comment (cry) [inline flask_core/core/templates/core/home.html]: Is this supposed to be merged in?

**Review by cry — DISMISSED** [review]

File: flask_core/app.py (hunk 7 above)
Comment (zainafzal08) [reply flask_core/app.py]: fuk u

File: flask_core/config.py (hunk 9 above)
Comment (zainafzal08) [reply flask_core/config.py]: done

File: flask_core/core/templates/core/home.html (hunk 10 above)
Comment (zainafzal08) [reply flask_core/core/templates/core/home.html]: lol no

**Review by cry — APPROVED** [review]
I hate your line spacing..
